Bluetooth: Classic: impl a2dp delay report and get config

MarkWangChinese opened the pull request:
implement delay report and get configuration

MarkWangChinese committed:
bluetooth: a2dp: implement delay report

implement delay report

Signed-off-by: Mark Wang <yichang.wang@nxp.com>

gzh-terry commented on subsys/bluetooth/host/classic/avdtp.c:
```diff
@@ -1146,80 +1325,197 @@ void (*rsp_handler[])(struct bt_avdtp *session, struct net_buf *buf, uint8_t msg
 		}
 
 		if (sigid != 0U && sigid <= BT_AVDTP_DELAYREPORT &&
 		    cmd_handler[sigid - 1U] != NULL) {
```
`cmd_handler` -> `rsp_handler` here, although not related to this patch.

MarkWangChinese replied:
I noticed it before too. Let me create one simple pr to fix it.

gzh-terry commented on include/zephyr/bluetooth/classic/a2dp.h:
```diff
@@ -321,6 +322,10 @@ struct bt_a2dp_codec_ie {
 struct bt_a2dp_codec_cfg {
+#ifdef CONFIG_BT_A2DP_DELAY_REPORT
+	/** the delay reporting configured state */
+	bool delay_report;
```
Wondering what's the relation between `delay_report` and the codec configuration?

BTW, I would recommend not to add `CONFIG_BT_A2DP_DELAY_REPORT` in this structure. Otherwise the upper layer has to add this macro as well.

We can just leave them to `false` if the delay report is not supported.

MarkWangChinese commented on include/zephyr/bluetooth/classic/a2dp.h:
```diff
@@ -578,6 +591,56 @@ struct bt_a2dp_cb {
+	 */
+	void (*get_config_rsp)(struct bt_a2dp_stream *stream, struct bt_a2dp_codec_cfg *codec_cfg,
+			       uint8_t rsp_err_code);
+#ifdef CONFIG_BT_A2DP_DELAY_REPORT
```
`CONFIG_BT_A2DP_DELAY_REPORT` is controlled by application, so the callbacks are not needed if application disable `CONFIG_BT_A2DP_DELAY_REPORT`. For example: `struct bt_conn_cb` use the same way.

gzh-terry replied:
> `CONFIG_BT_A2DP_DELAY_REPORT` is controlled by application, so the callbacks are not needed if application disable `CONFIG_BT_A2DP_DELAY_REPORT`. For example: `struct bt_conn_cb` use the same way.

The reason is fine. But at the App side, I would prefer to register a callback function that is never called, rather than use a macro to control the callback stucture, which look like:
```C
#ifdef CONFIG_BT_A2DP_DELAY_REPORT
#ifdef CONFIG_BT_A2DP_SOURCE
	static void delay_report_req() {}
#endif
#ifdef CONFIG_BT_A2DP_SINK
	static void delay_report_rsp() {}
#endif
static const struct bt_a2dp_cb a2dp_cb = {
	.connected = connected;
	.disconnected = disconnected;
	.config_req = config_req;
	.reconfig_req = reconfig_req;
...
#ifdef CONFIG_BT_A2DP_DELAY_REPORT
#ifdef CONFIG_BT_A2DP_SOURCE
	.delay_report_req = delay_report_req;
#endif
#ifdef CONFIG_BT_A2DP_SINK
	.delay_report_rsp = delay_report_rsp;
#endif
#endif
};
```

gzh-terry replied:
Less macros in this `.h` file will make it easier to develop an App, and cause very little increase in code size (consider that the implementation of the apis are controled by the configs)

MarkWangChinese committed:
bluetooth: a2dp: implement get_config

implement get configuration

Signed-off-by: Mark Wang <yichang.wang@nxp.com>
(cherry picked from commit 54cae2e75bcb7ee38ec4d7c24cef8835e3836a05)

gzh-terry commented on include/zephyr/bluetooth/classic/a2dp.h:
```diff
@@ -578,6 +591,56 @@ struct bt_a2dp_cb {
+	 *
+	 *  Called when the get configuration operation is completed.
+	 *
+	 *  @param[in] stream    Pointer to stream object.
```
```
	 *  @param[in] codec_cfg
```

MarkWangChinese replied:
Fixed

gzh-terry commented on subsys/bluetooth/host/classic/avdtp.c:
```diff
@@ -66,6 +76,11 @@ enum sep_state {
 
+static uint8_t avdtp_get_tid(struct bt_avdtp *session)
+{
+	return session->tid_sent++ % 16; /* Loop for 16 because only 4 bits */
```
`++tid_sent` since the name is sent?

Alternatively, `tid_next` might be more acccurate.

MarkWangChinese committed:
bluetooth: avdtp: check buf tail room

check the buf tail room before adding data to buf.

Signed-off-by: Mark Wang <yichang.wang@nxp.com>
(cherry picked from commit 65f8a9e03baf899684198e707170ad17817f7ce5)

MarkWangChinese committed:
bluetooth: a2dp: use avdtp struct to fill data to net buf

avdtp defines the data format by struct, use it to fill net buf.

Signed-off-by: Mark Wang <yichang.wang@nxp.com>
(cherry picked from commit a89f4eddac6f13f981f2ca0b94f07edaad104afa)

MarkWangChinese commented on subsys/bluetooth/host/classic/a2dp.c:
```diff
@@ -1000,6 +1219,39 @@ int bt_a2dp_stream_send(struct bt_a2dp_stream *stream, struct net_buf *buf, uint
 }
 #endif
 
+#if defined(CONFIG_BT_A2DP_SINK)
```
It seems that `#ifdef` and `#if defined` are used together in bluetooth codes. I prefer to change all as `#ifdef` in future pr.

lylezhu2012 reviewed (commented):
From the delay report procedure, the first delay report shall be sent immediately after the SEP has been configured. I did not find it.

MarkWangChinese committed:
bluetooth: a2dp: avoid clearing the cmd req

If the previous cmd req is triggered, the cmd fail to trigger again and
keep the previous cmd req.

Signed-off-by: Mark Wang <yichang.wang@nxp.com>
(cherry picked from commit f4197d51b62799748a75c97bd0b68778185ff293)

MarkWangChinese committed:
bluetooth: a2dp: remove meaningless if

The `a2dp` is got by `req` through CONTAINER_OF, then the if always is
true.

Signed-off-by: Mark Wang <yichang.wang@nxp.com>
(cherry picked from commit 24ef59d835c69a30fe8a8e6569405791532d9e0d)

MarkWangChinese committed:
bluetooth: avdtp: unify the avdtp error code based on Zephyr err code

get the avdtp error code based on Zephyr err code to make the codes more
clear.

Signed-off-by: Mark Wang <yichang.wang@nxp.com>
(cherry picked from commit 1d4d97fdb41791501eb63934f096bcc88cb2da81)

MarkWangChinese commented on subsys/bluetooth/host/classic/a2dp.c:
```diff
@@ -864,6 +951,25 @@ static int bt_a2dp_abort_cb(struct bt_avdtp_req *req, struct net_buf *buf)
 }
 
+#ifdef CONFIG_BT_A2DP_SINK
+#ifdef CONFIG_BT_A2DP_DELAY_REPORT
```
Fixed by new commit.

MarkWangChinese commented on subsys/bluetooth/host/classic/a2dp.c:
```diff
@@ -864,6 +951,25 @@ static int bt_a2dp_abort_cb(struct bt_avdtp_req *req, struct net_buf *buf)
+	return 0;
+}
+#endif
+#endif
```
Fixed by new commit.

MarkWangChinese commented on subsys/bluetooth/host/classic/a2dp.c:
```diff
@@ -1000,6 +1106,39 @@ int bt_a2dp_stream_send(struct bt_a2dp_stream *stream, struct net_buf *buf, uint
 #endif
 
+#if defined(CONFIG_BT_A2DP_SINK)
+#ifdef CONFIG_BT_A2DP_DELAY_REPORT
```
Fixed by new commit.

MarkWangChinese commented on subsys/bluetooth/host/classic/a2dp.c:
```diff
@@ -506,6 +531,49 @@ static int a2dp_abort_ind(struct bt_avdtp *session, struct bt_avdtp_sep *sep, ui
+		if (err != 0 && *errcode == 0) {
+			*errcode = BT_AVDTP_BAD_ACP_SEID;
+		}
+	}
```
Added the `delay` value to the callback.

MarkWangChinese commented on subsys/bluetooth/host/classic/a2dp.c:
```diff
@@ -785,12 +874,10 @@ typedef void (*bt_a2dp_done_cb)(struct bt_a2dp_stream *stream);
 
-	ep = CONTAINER_OF(a2dp->ctrl_param.sep, struct bt_a2dp_ep, sep);
-	if ((ep->stream == NULL) || (CTRL_REQ(req) != &a2dp->ctrl_param)) {
+	if (ep->stream == NULL) {
```
I use one standalone commit to do the change, and explain it in the commit.

MarkWangChinese commented on subsys/bluetooth/host/classic/avdtp.c:
```diff
@@ -1389,6 +1389,64 @@ static void avdtp_delay_report_rsp(struct bt_avdtp *session, struct net_buf *buf
+		err = -ENOTSUP;
+	} else {
+		rsp_buf = avdtp_create_pdu(BT_AVDTP_ACCEPT, BT_AVDTP_GET_CONFIGURATION, tid);
+		if (!rsp_buf) {
```
Done

MarkWangChinese commented on subsys/bluetooth/host/classic/avdtp.c:
```diff
@@ -2214,33 +2216,55 @@ static int avdtp_process_configure_command(struct bt_avdtp *session, uint8_t cmd
 
 	return avdtp_send_cmd(session, buf, &param->req);
+
+uref_and_return:
```
Done

MarkWangChinese committed:
bluetooth: a2dp: save configuration when cmd success

In the cmd result callback, save the configuration to stream if result is
success.

Signed-off-by: Mark Wang <yichang.wang@nxp.com>
(cherry picked from commit 9a9477fa0214f245a66e24a2edd90132293d5086)

lylezhu2012 commented on subsys/bluetooth/host/classic/avdtp.c:
```diff
@@ -2163,7 +2381,46 @@ int bt_avdtp_abort(struct bt_avdtp *session, struct bt_avdtp_ctrl_params *param)
+
+int bt_avdtp_get_configuration(struct bt_avdtp *session, struct bt_avdtp_ctrl_params *param)
+{
+	return bt_avdtp_ctrl(session, param, BT_AVDTP_GET_CONFIGURATION,
+			     AVDTP_CONFIGURED | AVDTP_OPENING | AVDTP_OPEN | AVDTP_STREAMING |
+			     AVDTP_CLOSING);
```
The get configuration command should fail if the stream is in Idle, closing or aborting state. So, I think the get config command should be avoided to be sent if the stream is in Idle, closing or aborting state.

lylezhu2012 commented on subsys/bluetooth/host/classic/avdtp.c:
```diff
@@ -1331,6 +1373,127 @@ static void avdtp_abort_rsp(struct bt_avdtp *session, struct net_buf *buf, uint8
+	}
+}
+
+static void avdtp_get_configuration_cmd(struct bt_avdtp *session, struct net_buf *buf, uint8_t tid)
```
The failure should be replied if the stream is in idle, closing or aborting state,

MarkWangChinese committed:
bluetooth: a2dp: remove the digital hardcode

use the sizeof of struct or variable; use macro to do the data shift.

Signed-off-by: Mark Wang <yichang.wang@nxp.com>
(cherry picked from commit d6d567355d08cefd1e11d88563fc300928af4972)

MarkWangChinese commented on subsys/bluetooth/host/classic/avdtp.c:
```diff
@@ -2052,27 +2241,55 @@ static int avdtp_process_configure_command(struct bt_avdtp *session, uint8_t cmd
+		cap = net_buf_add(buf, sizeof(*cap));
+		cap->service_category = BT_AVDTP_SERVICE_DELAY_REPORTING;
+		cap->losc = 0;
+	}
```
Done

MarkWangChinese commented on subsys/bluetooth/host/classic/avdtp.c:
```diff
@@ -1331,6 +1373,127 @@ static void avdtp_abort_rsp(struct bt_avdtp *session, struct net_buf *buf, uint8
+				   tid);
+	if (!rsp_buf) {
+		return;
+	}
```
done

MarkWangChinese commented on subsys/bluetooth/host/classic/avdtp.c:
```diff
@@ -2052,27 +2241,55 @@ static int avdtp_process_configure_command(struct bt_avdtp *session, uint8_t cmd
+	cap->losc = param->codec_specific_ie_len + 2;
+
+	media_cap = net_buf_add(buf, sizeof(*media_cap));
+	media_cap->media_type = param->media_type << 4U;
```
Updated to use macro.

MarkWangChinese committed:
bluetooth: avdtp: add copyright

add copyright to avdtp.c

Signed-off-by: Mark Wang <yichang.wang@nxp.com>
(cherry picked from commit 8761b2884726ccc33f953e4679a3d9c86b17da84)

lylezhu2012 reviewed (commented):
Since the delay report is a mandatory feature for A2DP sink, and the delay report is supported in this PR, it is not reasonable to add a separate commit to remove kconfig BT_A2DP_DELAY_REPORT and BT_AVDTP_DELAY_REPORT. My suggestion is unpacking the original commit (delay port support commit) and making the changes directly.

MarkWangChinese committed:
bluetooth: shell: a2dp: use static for functions

add static for functions and variables.

Signed-off-by: Mark Wang <yichang.wang@nxp.com>
(cherry picked from commit 1282bf7b8d814ef7a217c6f0c424aa4b2f34d363)

MarkWangChinese committed:
bluetooth: shell: a2dp: remove unused function

remove app_configured

Signed-off-by: Mark Wang <yichang.wang@nxp.com>
(cherry picked from commit c0be42f10341ea4a6ed1df152375335a5f514c92)

MarkWangChinese commented on subsys/bluetooth/host/classic/shell/a2dp.c:
```diff
@@ -801,6 +850,31 @@ static int cmd_send_media(const struct shell *sh, int32_t argc, char *argv[])
+static int cmd_send_delay_report(const struct shell *sh, int32_t argc, char *argv[])
+{
+	int err;
+
```
Done
